### ISSUE TITLE
Still race conditions when hitting server after reset

### ISSUE THREAD

**Issue by jnicklas** [jnicklas/capybara#1692]
The `wait_for_pending_requests` method on Server was introduced in 2.7.0, but unfortunately this method is not enough to prevent requests from hitting the server after reset. I have pretty definitive proof of this race condition occuring, and I think I have an explanation of what's happening:

- `wait_for_pending_requests` called, counter is zero, method returns
- request starts, hits server, count is increased to one
- `DatabaseCleaner.clean` is called, starts emptying out the DB
- request keeps processing, hits a point where a query fails due to missing data cleaned out by database cleaner, server sets `error` property
- next test starts, fails due to `error` property being set

In other words: requests can hit the application *in between* `wait_for_pending_requests` and `DatabaseCleaner.start`.

I managed to fashion a workaround for this issue, by introducing a request blocker, which prevents any requests from hitting the Rails application, and responds with a 200 status code and a "Request blocked" message: https://gist.github.com/jnicklas/090d7099afd6cb6349c5245e39d1e664

The order in which to apply the reset operations is very important though: the request blocker needs to be enabled *before* resetting sessions, otherwise this won't work. This way we prevent any new requests from hitting the application, as well as waiting for any existing requests to finish. I think given this setup, it is impossible for any requests to hit the application after session reset, and indeed after this change I cannot reproduce the race condition any more.

Unfortunately, integrating this into Capybara is not super straight-forward. Of course we could bundle something like the request block middleware attached in the gist, but it's a significantly breaking change. It would mean that people would have to call something like `session.start` before they can call any Capybara commands. For RSpec and Cucumber we could make this automatic, but it would be a breaking change for anyone using Capybara with another test framework or stand-alone.

Another option is to add some kind of reset hook, so that we could do something like:

``` ruby
Capybara.after_reset do
  DatabaseCleaner.start
end
```

Somewhere in the Capybara codebase we'd then do something like (pseudocode):

```
RequestBlocker.enable!
@before_reset_hooks.each(&:call)
server.wait_for_pending_requests
reset_session!
@after_reset_hooks.each(&:call)
RequestBlocker.disable!
```

This would actually be quite a nice feature, and I think we could do it without breaking backwards-compatibility. The problem is that it's somewhat unclear when to call these hooks. Capybara can have multiple sessions, and it's totally legit to reset one of those in the middle of a test. We don't want to go emptying out people's databases just because they reset a session in the middle of a test. Maybe that's just a question of API design, but I'm not sure.

**Comment by twalpole** [comment]
@jnicklas I think this may need more analysis -  The way it works currently - Session#reset! calls driver.reset!  which should be synchronous and not return until the browser has moved from the current page to about:blank (or equivalent).  There should be no way for the browser to initiate another request after this.  Then wait_for_pending_requests is called to wait for any orphaned requests still being processed by the app to finish.  Once they are done then reset! finishes.   If the new request is coming from a second session that should also be fine because all sessions get reset before the test ends and DatabaseCleaner starts up.  
If this is happening because someone is manually managing sessions and calling DatabaseCleaner after resetting only one session but keeping another session running there is nothing we can do about that, its user error

**Comment by twalpole** [comment]
@jnicklas Do you happen to be opening multiple windows in the session (and not closing them) where this is happening (as opposed to multiple sessions) ?  I think only the currently active window gets navigated to about: blank on driver#reset! (in selenium) which means you could potentially be getting a request initiated from another window.

**Comment by jnicklas** [comment]
I do have multiple sessions, but not multiple windows in this test. I can't really explain it either, and it's super hard to create a reproducible test case, but I'm *sure* that the app is indeed receiving requests in between `wait_for_pending` and `reset`. One possible factor is that we're using Puma as the server for Capybara, and not WebRICK. Could it be an explanation that Puma is "holding" a request? That is when we call `wait_for_pending_requests`, the request has already hit Puma, but Puma has not yet let it hit the application?

**Comment by jnicklas** [comment]
I think I actually managed to repro this! https://gist.github.com/jnicklas/754cc8ed42122f52b92abc7196c437e9

It's a bit weird, I'll admit and it took some fiddling with the parameters to get it to fail like this. The Queue is there to make sure that something like I/O flushing doesn't affect the result, the result is the same when using `puts` directly though.

As you can see in the sample output, we receive a request **after** calling reset, and after printing "done"! In theory this should never happen.

There are a bunch of conditions which seem to have to be met in order for this to fail:

- Server must be saturated with requests
- Multiple sessions must be used
- beforeunload hook which sends an AJAX request (maybe it just needs to be close in time to reset, but I couldn't get it to fail without beforeunload)

So all in all an unlikely scenario, but I think it proves that there is indeed a race condition in there.

**Comment by jnicklas** [comment]
I'm not sure about the repro though, it's really complicated, so please double check that I haven't done anything dumb.

**Comment by twalpole** [comment]
I guess puma could be holding a queue of requests -- hmmm -- - I'll take a look at the repro case in a bit

**Comment by jnicklas** [comment]
The repro doesn't use Puma, just webrick.

**Comment by twalpole** [comment]
Ok - playing with the repro.rb file --- If instead of `oReq.open("GET", "/ajax");` one does oReq.open("GET", "/ajax" + counter) - it works properly.  So I thought maybe it was timing so I added var constant = 10 and did oReq.open("GET", "/ajax" + constant) - still failed, and it fails even if counter is initially 1.  
  For some reason if the browser has already made a request to the URL - the  before_unload triggered request creates an issue - if it is a new URL everything works properly???  Is Firefox rate limiting requests to the same URL or something?

**Comment by jnicklas** [comment]
Yeah, this whole thing is super weird. It's kind of fascinating that this issue actually cropped up pretty consistently in a real app, given how hard it is to replicate. I get the same behaviour with Chrome btw, so I don't think it's anything browser specific.

**Comment by twalpole** [comment]
@jnicklas Found it - each session creates it's own instance of Server, but doesn't actually start it up when it's sharing between sessions, this means it actually ends up looking at the wrong request counter when trying to close the second session.  I thought I had made sure this case was taken care of, but I guess not.  A potential solution would be to just reset sessions in the reverse order they were created but they're currently stored in a hash so we don't have that info.  A short term fix for your use case would be to specify `Capybara.reuse_server = false`  which would fix the issue, but use a bit more memory.

**Comment by twalpole** [comment]
Actually hashes do iterate in insertion order (since ruby 1.9) so that may actually be a solution to this issue

**Comment by twalpole** [comment]
@jnicklas Please try PR #1693 and see if it fixes the issue on your real app (without Capybara.reuse_server = false).  If so it'll be a good enough fix for 2.7 and for 2.8 I think I'll look at making it so there is an instance of Counter per app instance and the sessions access the correct one.

**Comment by twalpole** [comment]
I've released this as 2.7.1 with the reversing of session resets which should fix this issue for now.  `Capybara.reuse_server = false` will also definitely fix the issue since each session would have a server tracking its requests.
